File: packages/runtime-native/tests/rg11b10-renderable.test.mjs
import assert from "node:assert/strict";
import { execFileSync } from "node:child_process";
import { existsSync } from "node:fs";
import { join } from "node:path";
import { fileURLToPath } from "node:url";
import { test } from "vitest";

const root = fileURLToPath(new URL("../", import.meta.url));

/**
 * `rg11b10ufloat-renderable` is the feature behind three's SSGI target format: SSGINode builds
 * its GI target as RGBFormat + UnsignedInt101111Type, the WebGPU backend maps that to
 * `rg11b10ufloat` with RENDER_ATTACHMENT usage, and rendering into it is gated on this optional
 * feature. The runtime never requested it and its JS feature-name mapper did not know the name,
 * so on native the stage could only appear as a device loss — which games then hid behind an
 * `isNative()` stage kill.
 *
 * Four layers must agree, and the test has all four: the raw Dawn probe (the oracle no binding
 * can lie to), context.cpp's adapter probe line, the JS adapter surface, and the JS device
 * surface. Agreement is what makes restoring the refusal a red rather than a skip. When the
 * adapter has the feature the contract also runs an actual render pass into the format — the
 * pre-fix device loss — and requires the device to survive it.
 *
 * The executable needs no display, so this belongs to the native-contract lane.
 */
/**
 * Adapters that are not a GPU. `Null backend` is Dawn's fallback when no Vulkan driver loads at
 * all; llvmpipe, lavapipe, softpipe and SwiftShader are CPU rasterisers. GitHub's runners have
 * no GPU, so this list is what separates "this machine cannot run the contract" from "the
 * bindings broke it".
 */
const SOFTWARE_ADAPTER = /Headless adapter: (.*(?:llvmpipe|lavapipe|softpipe|swiftshader|Null backend).*)/iu;

const ENGINE_BUILDS = [
  { engine: "V8", directory: "build/tn-linux" },
  { engine: "QuickJS", directory: "build/tn-linux-quickjs" },
];

test.each(ENGINE_BUILDS)(
  "rg11b10ufloat-renderable is truthful at every layer on $engine, and a pass into the format does not lose the device",
  ({ engine, directory }) => {
  const executable = join(root, directory, "threenative-rg11b10-renderable-test");
  if (!existsSync(executable)) {
    // Never a silent pass: an unbuilt executable is unexecuted, and says so.
    assert.fail(
      `${executable} is not built. Run: cmake --build ${directory} --target threenative-rg11b10-renderable-test`,
    );
  }
  let output;
  try {
    output = execFileSync(executable, { encoding: "utf8", stdio: ["ignore", "pipe", "pipe"] });
  } catch (error) {
    output = `${error.stdout ?? ""}${error.stderr ?? ""}`;
    const adapter = SOFTWARE_ADAPTER.exec(output);
    if (adapter) {
      console.info(
        `TN_RG11B10_UNEXECUTED: ${engine} on ${adapter[1].trim()} — no GPU adapter on this machine, so the contract did not execute.`,
      );
      return;
    }
    assert.fail(
      `${executable} failed on a hardware adapter, which is a real refusal:\n${output.slice(-2000)}`,
    );
  }
  assert.match(
    output,
    engine === "V8" ? /Creating V8 engine/u : /Creating QuickJS engine/u,
    `${directory} did not run under ${engine}`,
  );
  const rawProbe = /raw wgpuAdapterHasFeature: (yes|no)/u.exec(output);
  assert.ok(rawProbe, "the executable did not print its raw Dawn oracle result at all");
  const adapterProbe = /adapter feature probe rg11b10ufloat-renderable: (yes|no)/u.exec(output);
  assert.ok(
    adapterProbe,
    "context.cpp did not print its adapter probe for rg11b10ufloat-renderable",
  );
  assert.equal(
    adapterProbe[1],
    rawProbe[1],
    "context.cpp's adapter probe and the raw Dawn oracle disagree",
  );
  const reported = /TN_RG11B10:(\{.*\})/u.exec(output);
  assert.ok(reported, "the executable reported no rg11b10ufloat result at all");
  const result = JSON.parse(reported[1]);

  if (adapterProbe[1] === "no") {
    // This adapter genuinely lacks the feature. Recorded as unexecuted rather than passed.
    assert.equal(
      result.adapterHas,
      false,
      "bindings advertised a feature the adapter does not have",
    );
    console.info(
      "TN_RG11B10_UNEXECUTED: adapter lacks rg11b10ufloat-renderable, so the pass contract did not execute.",
    );
    return;
  }

  assert.equal(
    result.adapterHas,
    true,
    "the adapter advertises rg11b10ufloat-renderable but the JS adapter surface hid it",
  );
  assert.equal(
    result.deviceHas,
    true,
    "the adapter advertises rg11b10ufloat-renderable but the bindings refused it on the device",
  );
  assert.equal(
    result.passRan,
    true,
    "the rg11b10ufloat render pass never ran",
  );
});

File: packages/runtime-native/tests/rg11b10_renderable_test.cpp
// `rg11b10ufloat-renderable` must be truthful at the JS surface and usable by a render pass.
//
// Three's SSGINode builds its GI target as RGBFormat + UnsignedInt101111Type, which the WebGPU
// backend maps to `rg11b10ufloat` with RENDER_ATTACHMENT usage — unconditionally: no feature
// check, no reduced-quality fallback. Rendering into that format requires the optional WebGPU
// feature `rg11b10ufloat-renderable`. The runtime's device creation never requested it and the
// JS feature-name mapper did not know the name, so on native the stage could only manifest as a
// device loss — which lumen-hall's postprocessing.ts worked around by dropping GI on isNative().
//
// Four answers, from four independent layers:
//
//   raw Dawn   — wgpuAdapterHasFeature called here directly, the oracle no binding can lie to
//   probe line — the `[WebGPU] adapter feature probe` context.cpp prints at device creation
//   adapter JS — adapter.features.has("rg11b10ufloat-renderable")
//   device JS  — device.features.has("rg11b10ufloat-renderable")
//
// plus the thing the feature buys: a render pass into an rg11b10ufloat target that leaves the
// device alive. An adapter without the feature degrades to a reported unexecuted rather than a
// failure, because "this GPU cannot" and "these bindings cannot" are different results and only
// the second is a red.

#include "mystral/runtime.h"

#include <iostream>
#include <string>

// The same C API both backends implement; bindings.cpp reaches it the same way. The wgpu C
// header comes before webgpu_compat.h — the compat macros expand against names it declares.
#if defined(MYSTRAL_WEBGPU_DAWN)
#include <webgpu/webgpu.h>
#else
#include <webgpu.h>
#endif
#include "mystral/webgpu_compat.h"

namespace {

// Read by the JS through string injection: the oracle no binding can lie to. Everything else
// the JS reports about features is checked against this.
static bool g_rawAdapterHas = false;

constexpr const char* kSetup = R"JS((() => {
  const adapter = navigator.gpu.requestAdapter();
  const adapterHas = adapter.features.has("rg11b10ufloat-renderable");
  const device = adapter.requestDevice();
  const deviceHas = device.features.has("rg11b10ufloat-renderable");
  globalThis.__tnRg = { rawHas: RAW_ADAPTER_HAS, adapterHas, deviceHas, passRan: false };
  // Held off the reported record: JSON.stringify would otherwise serialise the whole device.
  globalThis.__tnRgDevice = device;
  if (!adapterHas) return;

  // The thing the feature buys. Before the fix this pass was the device loss itself: three's
  // SSGI creates exactly this target shape, Dawn validated RENDER_ATTACHMENT on a format the
  // device's feature set did not cover, and the run died instead of degrading.
  const target = device.createTexture({
    size: { width: 64, height: 64, depthOrArrayLayers: 1 },
    format: "rg11b10ufloat",
    usage: 0x10,                                                               // RENDER_ATTACHMENT
  });
  const view = device.createTextureView(target);
  const encoder = device.createCommandEncoder();
  const pass = encoder.beginRenderPass({
    colorAttachments: [
      { view, loadOp: "clear", storeOp: "store",
        clearValue: { r: 0.25, g: 0.5, b: 0.75, a: 1 } },
    ],
  });
  pass.end();
  device.queue.submit([encoder.finish()]);
  globalThis.__tnRg.passRan = true;
})())JS";

// Runs after the host has drained frames, so the pass above has actually crossed the packed
// frame op stream and been replayed against Dawn — the point where the pre-fix device loss
// surfaced. An empty submit plus onSubmittedWorkDone is the liveness probe: a device lost
// during the rg11b10 pass never completes this one.
constexpr const char* kRead = R"JS((() => {
  if (!globalThis.__tnRg.adapterHas) {
    console.log("TN_RG11B10:" + JSON.stringify({
      adapterHas: false, deviceHas: globalThis.__tnRg.deviceHas,
    }));
    return;
  }
  const device = globalThis.__tnRgDevice;
  const encoder = device.createCommandEncoder();
  device.queue.submit([encoder.finish()]);
  device.queue.onSubmittedWorkDone().then(() => {
    globalThis.__tnRgDone = true;
    console.log("TN_RG11B10:" + JSON.stringify(globalThis.__tnRg));
  }).catch((error) => {
    globalThis.__tnRgError = String(error && error.message ? error.message : error);
  });
})())JS";

constexpr const char* kAssert = R"JS((() => {
  const r = globalThis.__tnRg;
  if (!r.adapterHas) {
    // The adapter layer may honestly lack the feature. What it may not do is disagree with
    // the raw probe or grant the device something the adapter never had.
    if (r.rawHas) throw new Error("raw Dawn reports the feature but the JS adapter surface hid it");
    if (r.deviceHas) throw new Error("device was granted a feature the adapter does not advertise");
    console.log("TN_RG11B10_UNEXECUTED: adapter lacks rg11b10ufloat-renderable");
    return;
  }
  if (r.rawHas !== true) throw new Error("JS adapter advertised the feature raw Dawn does not have");
  if (r.deviceHas !== true)
    throw new Error("the adapter advertises rg11b10ufloat-renderable but the bindings refused it on the device");
  if (r.passRan !== true) throw new Error("setup never recorded the rg11b10ufloat render pass");
  if (globalThis.__tnRgError) throw new Error(globalThis.__tnRgError);
  if (globalThis.__tnRgDone !== true) throw new Error("liveness submit never completed");
})())JS";

#if defined(MYSTRAL_WEBGPU_DAWN)
struct RawAdapterData {
    WGPUAdapter adapter = nullptr;
    bool completed = false;
};

static void onRawAdapterRequest(WGPURequestAdapterStatus status, WGPUAdapter adapter,
                                WGPUStringView message, void* userdata1, void* userdata2) {
    auto* data = static_cast<RawAdapterData*>(userdata1);
    if (status == WGPURequestAdapterStatus_Success) data->adapter = adapter;
    else {
        std::cerr << "[RG11B10] raw adapter request failed: "
                  << (message.data ? std::string(message.data, message.length) : std::string())
                  << '\n';
    }
    data->completed = true;
}
#endif

}  // namespace

int main() {
#if defined(MYSTRAL_WEBGPU_DAWN)
    // The oracle: raw Dawn, no binding in between. This is what makes a refusal a red rather
    // than a skip — without it, a hidden feature and an absent feature look identical from JS.
    {
        WGPUInstanceDescriptor instanceDesc = {};
        WGPUInstance instance = wgpuCreateInstance(&instanceDesc);
        if (!instance) {
            std::cerr << "[RG11B10] could not create a raw Dawn instance for the oracle\n";
            return 1;
        }
        WGPURequestAdapterOptions adapterOptions = {};
        adapterOptions.powerPreference = WGPUPowerPreference_HighPerformance;
        RawAdapterData adapterData;
#if WGPU_USES_CALLBACK_INFO_PATTERN
        WGPURequestAdapterCallbackInfo callbackInfo = {};
        callbackInfo.mode = WGPUCallbackMode_AllowProcessEvents;
        callbackInfo.callback = onRawAdapterRequest;
        callbackInfo.userdata1 = &adapterData;
        callbackInfo.userdata2 = nullptr;
        wgpuInstanceRequestAdapter(instance, &adapterOptions, callbackInfo);
#else
        wgpuInstanceRequestAdapter(instance, &adapterOptions, onRawAdapterRequest, &adapterData);
#endif
        while (!adapterData.completed) {
            wgpuInstanceProcessEvents(instance);
        }
        if (!adapterData.adapter) {
            std::cerr << "[RG11B10] raw adapter request returned no adapter\n";
            return 1;
        }
        g_rawAdapterHas =
            wgpuAdapterHasFeature(adapterData.adapter, WGPUFeatureName_RG11B10UfloatRenderable) != 0;
        std::cout << "[RG11B10] raw wgpuAdapterHasFeature: " << (g_rawAdapterHas ? "yes" : "no")
                  << std::endl;
    }
#else
    std::cout << "[RG11B10] backend is not Dawn; raw oracle unavailable" << std::endl;
#endif

    mystral::RuntimeConfig config;
    config.width = 1;
    config.height = 1;
    config.noSdl = true;

    auto runtime = mystral::Runtime::create(config);
    if (!runtime) {
        std::cerr << "could not create headless native runtime\n";
        return 1;
    }

    // Inject the oracle's answer; the JS compares every layer it can see against it.
    const std::string setupWithOracle = [](const std::string& tpl) {
        const std::string needle = "RAW_ADAPTER_HAS";
        const std::string value = g_rawAdapterHas ? "true" : "false";
        std::string out = tpl;
        const auto at = out.find(needle);
        if (at != std::string::npos) out.replace(at, needle.size(), value);
        return out;
    }(std::string(kSetup));

    if (!runtime->evalScript(setupWithOracle.c_str(), "rg11b10_setup.js")) {
        std::cerr << "rg11b10ufloat bindings failed at setup";
        if (runtime->getExitCode() != 0) std::cerr << " (exit " << runtime->getExitCode() << ")";
        std::cerr << '\n';
        return 1;
    }

    // Frames, not ticks: the packed frame op stream is drained and replayed by the host once
    // per frame, and the render pass this contract is about sits in it. This is where the
    // pre-fix device loss happened.
    for (int frame = 0; frame < 60; frame += 1) {
        if (!runtime->pollEvents()) break;
    }

    if (!runtime->evalScript(kRead, "rg11b10_read.js")) {
        std::cerr << "rg11b10ufloat liveness probe failed to start\n";
        return 1;
    }
    for (int frame = 0; frame < 60; frame += 1) {
        if (!runtime->pollEvents()) break;
    }

    if (!runtime->evalScript(kAssert, "rg11b10_assert.js")) {
        std::cerr << "rg11b10ufloat contract failed\n";
        return 1;
    }

    std::cout << "native rg11b10 renderable bindings contract passed\n";
    return 0;
}
